### PR TITLE
Add initial DSPy reliability tests

### PR DESCRIPTION
**Pull request by dbczumar**
Add initial tests for DSPy reliability with real LLMs

### CODE REVIEW

File: .github/workflows/run_tests.yml [hunk 1]
```diff
@@ -62,7 +62,7 @@ jobs:
         with:
           args: check --fix-only
       - name: Run tests with pytest
-        run: poetry run pytest tests/
+        run: poetry run pytest tests/ --ignore=tests/quality
```
Comment (dbczumar) [inline .github/workflows/run_tests.yml:65]: Quality tests aren't ready to run on PRs yet

File: tests/quality/conftest.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    "gpt-o1-mini",
+    "claude-3.5-sonnet",
+    "claude-3.5-haiku",
+    "gemini-1.5-pro",
+    "gemini-1.5-flash",
+    "llama-3.1-405b-instruct",
+    "llama-3.1-70b-instruct",
+    "llama-3.1-8b-instruct",
+    "llama-3.2-3b-instruct",
+]
```
Comment (dbczumar) [inline tests/quality/conftest.py:24]: It's probably useful for us to define a canonical list of models that DSPy tests against. We can extend / update this list over time. For each model in this list, users can specify corresponding LiteLLM configurations in `quality_conf.yaml`, as described in the README

File: tests/quality/README.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            model: "anthropic/claude-3.5"
+            api_key: "<my_anthropic_api_key>"
+
+- Second, to run the tests, run the following command from this directory:
+
+  ```bash
+      pytest .
+  ```
+
+  This will execute all tests for the configured models and display detailed results for each model configuration. Tests are set up to mark expected failures for known challenging cases where a specific model might struggle, while actual (unexpected) DSPy quality issues are flagged as failures (see below).
```
Comment (dbczumar) [inline tests/quality/README.md:48]: As a follow-up, I'll write a pytest post suite hook that produces a report summarizing the quality results by model. Eventually, we can publish this report as we make DSPy releases

File: tests/quality/README.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+### Known Failing Models
+
+Some tests may be expected to fail with certain models, especially in challenging cases. These known failures are logged but do not affect the overall test result. This setup allows us to keep track of model-specific limitations without obstructing general test outcomes. Models that are known to fail a particular test case are specified using the `@known_failing_models` decorator. For example:
+
+```
+@known_failing_models(["llama-3.2-3b-instruct"])
```
Comment (dbczumar) [inline tests/quality/README.md:55]: As an alternative, I thought about having each test declare which models *pass* and only surfacing failures for model is in the *pass* list, but I rejected this because:

1. It becomes too easy to introduce a model with significant undetected failures that we might want to fix. If the default behavior for adding a new model is that all tests pass regardless of whether the model actually performs well, it's likely that we'll end up with many undisclosed model failures.

2. Declaring failures makes it much clearer where developers should spend time trying to make improvements. It's easier to see which items are present in a list (failing models present in a list of failing models), rather than which items are absent (failing models absent from a list of passing models).

File: tests/quality/README.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+### Running the Tests
+
+- First, populate the configuration file `quality_tests_conf.yaml` (located in this directory) with the necessary LiteLLM model/provider names and access credentials for 1. each LLM you want to test and 2. the LLM judge that you want to use for assessing the correctness of outputs in certain test cases. These should be placed in the `litellm_params` section for each model in the defined `model_list`. You can also use `litellm_params` to specify values for LLM hyperparameters like `temperature`. Any model that lacks configured `litellm_params` in the configuration file will be ignored during testing.
+
+  The configuration must also specify a DSPy adapter to use when testing, e.g. `"chat"` (for `dspy.ChatAdapter`) or `"json"` (for `dspy.JSONAdapter`)
+
+  An example of `quality_tests_conf.yaml`:
+
+      ```yaml
+      adapter: chat
```
Comment (dbczumar) [inline tests/quality/README.md:24]: Later, we can add cache configurations to this file

File: tests/quality/README.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+- **Cross-Model Compatibility**: By testing with different LLMs, we aim to ensure that DSPy programs perform well across model types and configurations, reducing model-specific edge cases and enhancing program versatility.
+
+### Running the Tests
+
+- First, populate the configuration file `quality_tests_conf.yaml` (located in this directory) with the necessary LiteLLM model/provider names and access credentials for 1. each LLM you want to test and 2. the LLM judge that you want to use for assessing the correctness of outputs in certain test cases. These should be placed in the `litellm_params` section for each model in the defined `model_list`. You can also use `litellm_params` to specify values for LLM hyperparameters like `temperature`. Any model that lacks configured `litellm_params` in the configuration file will be ignored during testing.
+
+  The configuration must also specify a DSPy adapter to use when testing, e.g. `"chat"` (for `dspy.ChatAdapter`) or `"json"` (for `dspy.JSONAdapter`)
+
+  An example of `quality_tests_conf.yaml`:
```
Comment (dbczumar) [inline tests/quality/README.md:21]: CI/CD systems (e.g. GitHub Actions workflows) can define their own test conf YAMLs for periodic regression testing

File: tests/quality/test_pydantic_models.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
+from enum import Enum
```
Comment (dbczumar) [inline tests/quality/test_pydantic_models.py:1]: I'll check in additional test cases once we confirm that the overall approach here makes sense

File: tests/quality/README.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
+# DSPy Quality Tests
```
Comment (dbczumar) [inline tests/quality/README.md:1]: Open to alternative names here. We could call this "Adapter" testing, but there's more being tested here than just Adapter logic - we're writing full programs and making quality assertions on their outputs.

I chose "quality" because one objective of DSPy is to provide consistent quality across LLMs

File: tests/quality/README.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+- **Challenging and Adversarial Tests**: Some of the tests are intentionally challenging or adversarial, crafted to push the boundaries of DSPy. These challenging cases allow us to gauge the robustness of DSPy and identify areas for potential improvement.
+- **Cross-Model Compatibility**: By testing with different LLMs, we aim to ensure that DSPy programs perform well across model types and configurations, reducing model-specific edge cases and enhancing program versatility.
+
+### Running the Tests
+
+- First, populate the configuration file `quality_tests_conf.yaml` (located in this directory) with the necessary LiteLLM model/provider names and access credentials for 1. each LLM you want to test and 2. the LLM judge that you want to use for assessing the correctness of outputs in certain test cases. These should be placed in the `litellm_params` section for each model in the defined `model_list`. You can also use `litellm_params` to specify values for LLM hyperparameters like `temperature`. Any model that lacks configured `litellm_params` in the configuration file will be ignored during testing.
```
Comment (dbczumar) [inline tests/quality/README.md:17]: At some point, it seems plausible for us to define recommended hyperparameter configurations for specific LMs that are known to produce better performance (e.g. temperature)
